liquibase/liquibase
# [LB-1373] fixed precision for time type for PostgreSQL

## Environment
**Liquibase Version**: 4.3.5

**Liquibase Integration & Version**: 

**Liquibase Extension(s) & Version**: 

**Database Vendor & Version**: postgres 9-13

**Operating System Type & Version**: MacOS Big Sur 11.4

## Pull Request Type
closes #1774
[https://datical.atlassian.net/browse/LB-1373](https://datical.atlassian.net/browse/LB-1373)

* \[x] Bug fix (non-breaking change which fixes an issue.)
* \[ ] Enhancement/New feature (non-breaking change which adds functionality)
* \[ ] Breaking change (fix or feature that would cause existing functionality to change)

## Description
Precision in changeSet for Time Datatype was missing for PostreSQL in generated query (and in DB column properties respectfully)

## Steps To Reproduce
create xml changelog with next changeset 

```java
<changeSet id="1" author="any">
        <createTable tableName="datatypes_test_table">
            <column name="time" type="time"/>
            <column name="timeewtz" type="time with time zone"/>
            <column name="time0" type="time (0)"/>
            <column name="timee0wtz" type="time (0) with time zone"/>
            <column name="time1" type="time (1)"/>
            <column name="time1wtz" type="time (1) with time zone"/>
            <column name="time2" type="time (2)"/>
            <column name="time2wtz" type="time (2) with time zone"/>
            <column name="time3" type="time (3)"/>
            <column name="time3wtz" type="time (3) with time zone"/>
            <column name="time4" type="time (4)"/>
            <column name="time4wtz" type="time (4) with time zone"/>
            <column name="time5" type="time (5)"/>
            <column name="time5wtz" type="time (5) with time zone"/>
            <column name="time6" type="time (6)"/>
            <column name="time6wtz" type="time (6) with time zone"/> 
        </createTable>
    </changeSet>
```

Run Liquibase `updateSQL` command to observe SQL generated for changeLog, notice precision for time is missing

```java
CREATE TABLE datatypes_test_table
(
    time      TIME WITHOUT TIME ZONE,
    timeewtz  TIME WITH TIME ZONE,
    time0     TIME WITHOUT TIME ZONE,
    timee0wtz TIME WITH TIME ZONE,
    time1     TIME WITHOUT TIME ZONE,
    time1wtz  TIME WITH TIME ZONE,
    time2     TIME WITHOUT TIME ZONE,
    time2wtz  TIME WITH TIME ZONE,
    time3     TIME WITHOUT TIME ZONE,
    time3wtz  TIME WITH TIME ZONE,
    time4     TIME WITHOUT TIME ZONE,
    time4wtz  TIME WITH TIME ZONE,
    time5     TIME WITHOUT TIME ZONE,
    time5wtz  TIME WITH TIME ZONE,
    time6     TIME WITHOUT TIME ZONE,
    time6wtz  TIME WITH TIME ZONE
)
```

## Actual Behavior
for mentioned changeSet Liquibase generated query like this

```java
CREATE TABLE datatypes_test_table
(
    time      TIME WITHOUT TIME ZONE,
    timeewtz  TIME WITH TIME ZONE,
    time0     TIME WITHOUT TIME ZONE,
    timee0wtz TIME WITH TIME ZONE,
    time1     TIME WITHOUT TIME ZONE,
    time1wtz  TIME WITH TIME ZONE,
    time2     TIME WITHOUT TIME ZONE,
    time2wtz  TIME WITH TIME ZONE,
    time3     TIME WITHOUT TIME ZONE,
    time3wtz  TIME WITH TIME ZONE,
    time4     TIME WITHOUT TIME ZONE,
    time4wtz  TIME WITH TIME ZONE,
    time5     TIME WITHOUT TIME ZONE,
    time5wtz  TIME WITH TIME ZONE,
    time6     TIME WITHOUT TIME ZONE,
    time6wtz  TIME WITH TIME ZONE
)
```

with missing precision argument

## Expected/Desired Behavior
With this fix for mentioned changeSet Liquibase generates query like this

```java
CREATE TABLE datatypes_test_table
(
    time      time WITHOUT TIME ZONE,
    timeewtz  time WITH TIME ZONE,
    time0     time(0) WITHOUT TIME ZONE,
    timee0wtz time(0) WITH TIME ZONE,
    time1     time(1) WITHOUT TIME ZONE,
    time1wtz  time(1) WITH TIME ZONE,
    time2     time(2) WITHOUT TIME ZONE,
    time2wtz  time(2) WITH TIME ZONE,
    time3     time(3) WITHOUT TIME ZONE,
    time3wtz  time(3) WITH TIME ZONE,
    time4     time(4) WITHOUT TIME ZONE,
    time4wtz  time(4) WITH TIME ZONE,
    time5     time(5) WITHOUT TIME ZONE,
    time5wtz  time(5) WITH TIME ZONE,
    time6     time(6) WITHOUT TIME ZONE,
    time6wtz  time(6) WITH TIME ZONE
)
```

Run changeSet against PostgreSQL DB and check column properties - it also shows precision 

## Test Requirements (Liquibase Internal QA)
Prior to manual testing:

* Verify the Liquibase Funcitonal tests ran and passed as part of the Github Actions build. 
* Execute the liquibase/liquibase-test-harness test for [liquibase-test-harness](https://github.com/liquibase/liquibase-test-harness)/[src](https://github.com/liquibase/liquibase-test-harness/tree/main/src)/[main](https://github.com/liquibase/liquibase-test-harness/tree/main/src/main)/[resources](https://github.com/liquibase/liquibase-test-harness/tree/main/src/main/resources)/[liquibase](https://github.com/liquibase/liquibase-test-harness/tree/main/src/main/resources/liquibase)/[harness](https://github.com/liquibase/liquibase-test-harness/tree/main/src/main/resources/liquibase/harness)/[change](https://github.com/liquibase/liquibase-test-harness/tree/main/src/main/resources/liquibase/harness/change)/[changelogs](https://github.com/liquibase/liquibase-test-harness/tree/main/src/main/resources/liquibase/harness/change/changelogs)/[postgresql](https://github.com/liquibase/liquibase-test-harness/tree/main/src/main/resources/liquibase/harness/change/changelogs/postgresql)/*datatypes.datetime.xml* against:
  * Postgres 9
  * Postgres 9.5
  * Postgres 10
  * Postgres 11
  * Postgres 12
  * Postgres 13

**Manual Tests**

The manual tests need only be executed against one version of Postgres.

_Verify update-sql correctly generates Postgres SQL to create columns of datatype TIMEZONE with varying precision._

`liquibase update-sql --changelog-file lb1798-changelog.xml`

* Columns have the expected precision on the `time` datatype; ie `time(3)`.
* Columns have the expected TIMEZONE setting; ie `WITHOUT TIME ZONE`

_Verify a successful update of a Postgres database to create a table with TIMEZONE and precision datatype columns._

`liquibase update --changelog-file lb1798-changelog.xml`

* There are no errors during update.
* The table is created on the database.
  * Columns have the expected precision on the `time` datatype; ie `time(3)`.
  * Columns have the expected TIMEZONE setting; ie `WITHOUT TIME ZONE`

_Verify generate-changelog includes a createTable changeset with TIMEZONE datatype columns._

`liquibase generate-changelog --changelog-file gcl-1798-changelog.xml`

* Columns have the expected precision on the `time` datatype; ie `time(3)`.
* Columns have the expected TIMEZONE setting; ie `WITHOUT TIME ZONE`

**Automated Tests**

* No new functional automated tests are required for this ticket.
* Test harness test already created.
* Dev also added a unit test.

## Fast Track PR Acceptance Checklist:
* \[x] Build is successful and all new and existing tests pass
* \[x] Added [Unit Test(s)]
* \[x] Added [Test Harness Test(s)] - Added previously by Robert, test-harness helped to find this bug.



┆Issue is synchronized with this [Jira Bug](https://datical.atlassian.net/browse/LB-1798) by [Unito](https://www.unito.io)
